status-im/status-react
# Refactored message data-model and view

[comment]: # (Please replace ... with your information. Remove < and >)
[comment]: # (To auto-close issue on merge, please insert the related issue number after # i.e fixes #566)

addresses #1875 #2602 #2374 

### Summary:

[comment]: # (Summarise the problem and how the pull request solves it)
* Refactored chat data model to support random access to any chat message
* Removal of redundant data in data-model (`:message-data`)
* Simpler chat views with less logic
* More effective subscriptions (most importantly those related to message stream)
* Way more scalable count of active subscriptions in message stream

To illustrate the later point, in chat with ~50 messages, the subscription count is 250 in current develop versus 60 in this branch and what's the most important, it doesn't grow linearly.  

Furthermore, it would be quite easy to get rid of some other data which are logically tied to messages, such as requests, those don't need to be distinct entities at all, just `:answered?` flag for message of type `content-type-command-request` is totally sufficient, this will lower the subscription count event more and simplify view logic.

As the message stream is now captured as a subscription hooked on combination of 2 functions, we can unit test it very well (message ordering, logical clock, datemark positioning, etc.), something which was not possible before, when stream generation was directly in view (+ with strange hardcoded logic because of the console chat - again).
I didn't have the time to write such tests yet, I will do it very soon.

I didn't totally convert all the chat view namespaces into refactored ones (with defview, letsubs, no `:refer`-ing, etc) + I limited the changes to events namespaces to keep the LOC changed smaller.

### Steps to test:
- Open Status
- Throughly test all type of chats, pay attention to message grouping/ordering, delivery statuses and command previews, everything should work correctly
- Test upgrade paths for all type of chats and commands (upgrading from `0.9.12` with various chats (1-1, group, public) and in those chats, messages including all commands)



status: ready
